src/main/java/org/mockito/internal/debugging/Java9PlusLocationImpl.java

```diff
@@ -0,0 +1,262 @@
+
+    private static final StackTraceCleaner CLEANER =
+            Plugins.getStackTraceCleanerProvider()
+                    .getStackTraceCleaner(new DefaultStackTraceCleaner());
+
+    private static final Function<Object, StackFrameMetadata> toStackFrameMetadata =
+            MetadataShim::new;
+    private static final Predicate<StackFrameMetadata> cleanerIsIn = CLEANER::isIn;
+
+    private static final int FRAMES_TO_SKIP = framesToSkip();
+
+    private final StackFrameMetadata sfm;
+    private volatile String stackTraceLine;
+
+    Java9PlusLocationImpl(boolean isInline) {
+        this.sfm = getStackFrame(isInline);
+    }
+
+    @Override
+    public String getSourceFile() {
+        return sfm.getFileName();
+    }
+
+    @Override
+    public String toString() {
+        maybeInitializeStackTraceLine();
+        return stackTraceLine;
+    }
+
+    private void maybeInitializeStackTraceLine() {
+        if (stackTraceLine == null) {
+            synchronized (this) {
+                if (stackTraceLine == null) {
+                    stackTraceLine = stackTraceLine(sfm);
+                }
+            }
+        }
+    }
+
+    private static String stackTraceLine(StackFrameMetadata element) {
+        return PREFIX + element.toString();
+    }
+
+    private static StackFrameMetadata getStackFrame(boolean isInline) {
+        return stackWalk(
+                stream ->
+                        stream.map(toStackFrameMetadata)
+                                .skip(FRAMES_TO_SKIP)
+                                .filter(cleanerIsIn)
+                                .skip(isInline ? 1 : 0)
+                                .findFirst()
+                                .orElseThrow(
+                                        () ->
+                                                new IllegalArgumentException(
+                                                        "Mockito could not find the first non-Mockito stack frame")));
+    }
+
+    /**
+     * In order to trigger the stack walker, we create some reflective frames. These need to be skipped so as to
+     * ensure there are no non-Mockito frames at the top of the stack trace.
+     */
+    private static int framesToSkip() {
+        return stackWalk(
+                stream -> {
+                    List<String> metadata =
+                            stream.map(toStackFrameMetadata)
+                                    .map(StackFrameMetadata::getClassName)
+                                    .collect(Collectors.toList());
+                    return metadata.indexOf(Java9PlusLocationImpl.class.getName());
+                });
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T stackWalk(Function<Stream<Object>, T> function) {
+        try {
+            return (T) walk.invoke(stackWalker, function);
+        } catch (InvocationTargetException | IllegalAccessException e) {
+            throw new RuntimeException(e);
```

Comment:
Rather than `RuntimeException`, let's use `MockitoException` with a descriptive error message that we don't intend to ever reach this codepath. https://github.com/mockito/mockito/blob/main/src/main/java/org/mockito/internal/creation/bytebuddy/SubclassByteBuddyMockMaker.java has numerous examples of descriptive exceptions that would be helpful for the user. If the message ends up being the same for all of these, we can also define a custom subclass of `MockitoException` in this class and use that.

Reply:
I've made some error message changes which are hopefully helpful. If you would like the wording tweaked, please feel free to change the PR directly? These kinds of communications are usually some of the hardest bits to fit with the 'style' of the project.

Reply:
The new error messages are fine by me, thanks!